### DISCUSSION TITLE
Restoring proxy to a previous state

### DISCUSSION THREAD

**Discussion by alexander-entin** [pmndrs/valtio#834] [Q&A]
While debugging proxy state restoration I stumbled upon somewhat strange behavior.

````javascript
const initial = { x: 1 }
const p = proxy()
p.a = initial
p.a.x = 2
p.a = initial
console.log(p.a.x)
````
<details><summary>result</summary>
2
</details>

````javascript
const v = { x: 1 }
const p = proxy()
p.a = v
v.x++
p.b = p.a
p.c = v
p.c.x += 10
p.a = v
console.log(p.a.x, p.b.x, p.c.x)
````
<details><summary>result</summary>
11 11 11
</details>

My use case - snapshot restoration - worked as expected in a snippet, by does nothing in a real app as the value in question is present in `proxyCache` for some reason. I'm about to `deepClone` a snapshot before trying to restore it to disable the cache.

And my questions:

1. What is a recommended way to restore a proxy to the initial value or an intermediate snapshot?

2. Is the behavior demonstrated by the above snippets by design or a bug?

Thank you!

**Comment by dai-shi** [comment]
> What is a recommended way to restore a proxy to the initial value or an intermediate snapshot?

You need to deep clone.

> Is the behavior demonstrated by the above snippets by design or a bug?

It's by design. We want to support circular object structure and it's kind of a side effect.
Basically, we recommend not to reuse the initial object.
v2 #810 will change the behavior, but I don't think it helps your case of restoring.
